### ISSUE TITLE
Multiclass accuracy with micro and top-k does not work as expected.

### ISSUE THREAD

**Issue by mmcdermott** [Lightning-AI/torchmetrics#3068]
## 🐛 Bug

I expect that `top_k` should impact accuracy, even in micro mode -- namely, in that `top_k` should mean that you get a prediction "correct" if the "right" logit is among the "top `top_k`" scores, then you average the number of corrects as per usual. This is not the behavior I see with torchmetrics now. Instead, I see this:

```python
>>> logits
tensor([[[0.0000, 0.1000, 0.5000, 0.4000],
         [0.0000, 0.2000, 0.7000, 0.1000]],

        [[0.0000, 0.4000, 0.3000, 0.3000],
         [1.0000, 0.0000, 0.0000, 0.0000]]])
>>> code
tensor([[3, 2],
        [1, 0]])
>>> logits.shape
torch.Size([2, 2, 4])
>>> code.shape
torch.Size([2, 2])
>>> acc = Accuracy(task="multiclass", ignore_index=0, num_classes=4, multidim_average="global", average="micro", top_k=4)
>>> acc(logits.transpose(2, 1), code)
tensor(0.6667)
>>> acc = Accuracy(task="multiclass", ignore_index=0, num_classes=4, multidim_average="global", average="micro", top_k=3)
>>> acc(logits.transpose(2, 1), code)
tensor(0.6667)
>>> acc = Accuracy(task="multiclass", ignore_index=0, num_classes=4, multidim_average="global", average="micro", top_k=2)
>>> acc(logits.transpose(2, 1), code)
tensor(0.6667)
>>> acc = Accuracy(task="multiclass", ignore_index=0, num_classes=4, multidim_average="global", average="micro", top_k=1)
>>> acc(logits.transpose(2, 1), code)
tensor(0.6667)
```

### Expected behavior

I would expect to see all accuracies with top_k >= 2 to be 1.0.

### Additional context

This line https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/functional/classification/accuracy.py#L85 ignores `top_k` in the micro case.

I think the fix to this issue may be related: https://github.com/Lightning-AI/torchmetrics/issues/3037

**Comment by mmcdermott** [comment]
Also related: https://github.com/Lightning-AI/torchmetrics/issues/2418

**Comment by mmcdermott** [comment]
I think it is actually the case that it doesn't work as expected in any `multiclass` setting. It may be working as the developers intended, and just an issue with my expectations, but I would expect top-k in multiclass to consider a prediction correct if any of the top-k predictions were correct, then average those # of corrects. I'm not really sure what it is doing now.

**Comment by mmcdermott** [comment]
1.7.1
